apps/web-client/src/lib/db/cr-sqlite/order-reconciliation.ts

```diff
@@ -202,3 +203,61 @@ export async function finalizeReconciliationOrder(db: DB, id: number) {
 		}
 	});
 }
+
+/**
+* Processes delivered books against placed order lines to identify matches an
+discrepancies
+*
+* @param scannedBooks - Array of scanned books with their quantities
+* @param placedOrderLines - Array of originally placed order lines from
+supplier
+*
+* @returns Object containing:
+*  - processedLines: Array of matched books with both ordered and delivered
+quantities
+*  - unmatchedBooks: Array of books that either:
+*    - Were ordered but not delivered
+*    - Were delivered but not in original order
+*/
+export const processOrderDelivery = (
+	scannedBooks: (BookEntry & { quantity: number })[],
+	placedOrderLines: SupplierPlacedOrderLine[]
+): { processedLines: ProcessedOrderLine[]; unmatchedBooks: (BookEntry & { quantity: number })[] } => {
+	const scannedLinesMap = new Map<string, BookEntry & { quantity: number }>(
+		scannedBooks.map((book) => {
+			return [book.isbn, book];
+		})
+	);
+	const result = { processedLines: [], unmatchedBooks: [] };
+
+	for (const placedOrderLine of placedOrderLines) {
+		if (scannedLinesMap.has(placedOrderLine.isbn)) {
```

Comment:
This is looking good. However, I had a click through and noticed one case which isn't handled as I would have expected: an "under delivered" quantity of 0 where there is an ordered quantity (In the mock data you have set up isbn 123456789 appears in the "unmatched books" table). I think in this case the row should be reported in its supplier order group as Ordered Quantity = X, Delivered Quantity = 0. To handle this, it might be better to start by iterating scannedOrderLines & check against a placedOrderLines map?

Reply:
I could also add the placedOrderLine to the processedLines array (instead of the scannedBook to keep things consistent) and in the else block include a deliveredQuantity of 0

Reply:
Not sure I follow... the else block would need to decide if its pushing to processLines or unmatchedBooks?

Reply:
take a look at my latest commit

Reply:
Ok, makes sense. Only other two things I'd say:
- add some space. Lines 235 - 237 could use it, otherwise its hard to read the `scannedLinesMap.delete()` action, which is pretty important
- line 250:  `result.unmatchedBooks` doesn't need to be spread.